### PR TITLE
Respecting dependencies when inserting a cache blocking group into the run order for the OpenMP backend

### PR DESCRIPTION
**Pull request by TobyFlynn**
This pull request inserts a cache blocking group into a graph's run order while respecting dependencies. Previously, a group replaced its last running kernel in a graph's run order without checking for dependencies. Now, the position of each kernel within a group is considered, with the earliest position that doesn't violate any dependencies being selected.

I've also added a quick sanity check afterwards to check that this positioning does not violate other dependencies.

### CODE REVIEW

File: pyfr/backends/openmp/types.py [hunk 1]
```diff
@@ -182,7 +182,22 @@ def group(self, kerns, subs=[]):
             groups.append(rargs)
 
         # Arrange for the groupings to be inserted into the final run list
-        self.kins[max(self.knodes[k] for k in kerns) - 1] = groups
+        group_deps = [dep for k in kerns for dep in self.kdeps[k] if dep not in kerns]
+        group_kins_ind = None
+        if group_deps:
+            max_dep_ind = max(self.knodes[dep] for dep in group_deps)
+            group_kins_ind = min(self.knodes[k] for k in kerns if self.knodes[k] > max_dep_ind) - 1
```
Comment (FreddieWitherden) [inline pyfr/backends/openmp/types.py:189]: Line length; maybe use shorter variable names.

File: pyfr/backends/openmp/types.py [hunk 2]
```diff
@@ -182,7 +182,22 @@ def group(self, kerns, subs=[]):
...
+            self.kins[group_kins_ind] = groups
+        else:
+            group_kins_ind = min(self.knodes[k] for k in kerns) - 1
+            self.kins[group_kins_ind] = groups
```
Comment (FreddieWitherden) [inline pyfr/backends/openmp/types.py:193]: Can move this outside the block

File: pyfr/backends/openmp/types.py [hunk 3]
```diff
@@ -182,7 +182,23 @@ def group(self, kerns, subs=[]):
             groups.append(rargs)
 
         # Arrange for the groupings to be inserted into the final run list
-        self.kins[max(self.knodes[k] for k in kerns) - 1] = groups
+        gdeps = []
+        for k in kerns:
+            gdeps = gdeps + [dep for dep in self.kdeps[k] if dep not in kerns]
+        gid = None
```
Comment (FreddieWitherden) [inline pyfr/backends/openmp/types.py:188]: Why is this needed?  gid is assigned in both branches of the loop.

File: pyfr/backends/openmp/types.py [hunk 4]
```diff
@@ -182,7 +182,23 @@ def group(self, kerns, subs=[]):
             groups.append(rargs)
 
         # Arrange for the groupings to be inserted into the final run list
-        self.kins[max(self.knodes[k] for k in kerns) - 1] = groups
+        gdeps = []
```
Comment (FreddieWitherden) [inline pyfr/backends/openmp/types.py:185]: Use a list comprehension here.

File: pyfr/backends/openmp/types.py [hunk 5]
```diff
@@ -182,7 +182,23 @@ def group(self, kerns, subs=[]):
...
+        for k in kerns:
+            gdeps = gdeps + [dep for dep in self.kdeps[k] if dep not in kerns]
+        gid = None
+        if gdeps:
+            lk = max(self.knodes[dep] for dep in gdeps)
+            gid = min(self.knodes[k] for k in kerns if self.knodes[k] > lk) - 1
+        else:
+            gid = min(self.knodes[k] for k in kerns) - 1
+        self.kins[gid] = groups
```
Comment (FreddieWitherden) [inline pyfr/backends/openmp/types.py:194]: Add newline.

File: pyfr/backends/openmp/types.py (hunk 4 above)
Comment (TobyFlynn) [reply pyfr/backends/openmp/types.py:185]: That was what I was doing previously with `[dep for k in kerns for dep in self.kdeps[k] if dep not in kerns]`

Comment (FreddieWitherden) [reply pyfr/backends/openmp/types.py:185]: Should just need a line break or two.

File: pyfr/backends/openmp/types.py (hunk 3 above)
Comment (TobyFlynn) [reply pyfr/backends/openmp/types.py:188]: Oops I've very used to pre-declaring variables from other languages sorry

**Comment by FreddieWitherden** [comment]
May be able to unify everything by doing
```python
lk = max((self.knodes[dep] for dep in gdeps), default=-1)
```
which avoids the if completely.

**Comment by TobyFlynn** [comment]
> May be able to unify everything by doing
> 
> ```python
> lk = max((self.knodes[dep] for dep in gdeps), default=-1)
> ```
> 
> which avoids the if completely.

Nice!

File: pyfr/backends/openmp/types.py (hunk 5 above)
Comment (FreddieWitherden) [reply pyfr/backends/openmp/types.py:189]: Avoid `\`, they should almost never be used.  Newlines are allowed inside [] and ().

Comment (FreddieWitherden) [reply pyfr/backends/openmp/types.py:189]: Also, we might want to move down the `self.kins[gid] = groups` bit until after the sanity check.

File: pyfr/backends/openmp/types.py [hunk 6]
```diff
@@ -182,7 +182,18 @@ def group(self, kerns, subs=[]):
...
+        lk = max((self.knodes[dep] for dep in gdeps), default=-1)
+        gid = min(self.knodes[k] for k in kerns if self.knodes[k] > lk) - 1
+        self.kins[gid] = groups
+
+        # Sanity check that other dependencies haven't been violated
+        for k in self.kdeps:
+            if k not in kerns:
+                for dep in self.kdeps[k]:
+                    if dep in kerns and self.knodes[k] < gid:
+                        raise RuntimeError('Graph grouping dependency error')
```
Comment (FreddieWitherden) [inline pyfr/backends/openmp/types.py:196]: Line length.

File: pyfr/backends/openmp/types.py (hunk 5 above)
Comment (TobyFlynn) [reply pyfr/backends/openmp/types.py]: I think I've now fixed this

File: pyfr/backends/openmp/types.py [hunk 7]
```diff
@@ -182,7 +182,19 @@ def group(self, kerns, subs=[]):
             groups.append(rargs)
 
         # Arrange for the groupings to be inserted into the final run list
-        self.kins[max(self.knodes[k] for k in kerns) - 1] = groups
+        gdeps = [dep for k in kerns
+                        for dep in self.kdeps[k] if dep not in kerns]
```
[the remark this answers is not in the document]
Comment (TobyFlynn) [reply pyfr/backends/openmp/types.py:186]: No indent at all like this?
```
gdeps = [dep for k in kerns 
for dep in self.kdeps[k] if dep not in kerns]
```
or so it lines up after the assignment?
```
gdeps = [dep for k in kerns 
         for dep in self.kdeps[k] if dep not in kerns]
```

[the remark this answers is not in the document]
Comment (FreddieWitherden) [reply pyfr/backends/openmp/types.py:186]: The second form
